pkg/multicluster/clientset.go

```diff
@@ -0,0 +1,81 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rotisserie/eris"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+// ClientSet provides access to a client.Client for any registered cluster.
+// Only one ClientSet should be created within a multicluster system.
+type ClientSet interface {
+	// Cluster returns a client.Client for the given cluster if one is available, else errors.
+	Cluster(name string) (client.Client, error)
+	getManager(cluster string) (manager.Manager, error)
+	setManager(cluster string, mgr manager.Manager, cancel context.CancelFunc)
+	deleteManager(cluster string)
```

Comment:
what is the purpose of requiring private methods on a public interface?

Reply:
The private methods prevent users from implementing the interface. The implementation in this package is used by the ClusterWatcher to store the canonical set of managers per cluster. Thus, any other implementation of the interface that just exposes the `Cluster` function would not be compatible with the watcher.

This approach was inspired by https://github.com/golang/lint/issues/210#issuecomment-245736660.

Alternatively, I could have just used a struct with some private/public functions. That would be simpler.

Reply:
In the latest iteration, I split this interface into two:

```go
type ManagerSet interface {
	getManager(cluster string) (manager.Manager, error)
	setManager(cluster string, mgr manager.Manager, cancel context.CancelFunc)
	deleteManager(cluster string)
}
```
and
```go
type Client interface {
	// Cluster returns a client.Client for the given cluster if one is available, else errors.
	Cluster(name string) (client.Client, error)
}
```

Client and the clusterWatcher both depend on ManagerSet. Client uses it in order to build clients from an existing cluster manager, and the clusterWatcher adds/removes managers from it as kubeconfig secrets are added and removed.

I chose this approach because it hides internals from the user by preventing them from manually adding and removing managers to the set without hiding that the manager set is the canonical source of per-cluster managers. That said, I don't love the setup here, and would appreciate any additional feedback.